OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs

```diff
@@ -900,6 +900,53 @@ internal static void UpgradeWeaponRules(ModData modData, int engineVersion, ref
 					}
 				}
 
+				// Valid-/InvalidImpactTypes were removed from CreateEffectWarhead, which uses Valid-/InvalidTargets instead now
+				if (engineVersion < 20170625)
+				{
+					if (node.Key.StartsWith("Warhead", StringComparison.Ordinal) && node.Value.Value == "CreateEffect")
+					{
+						var validImpactTypes = node.Value.Nodes.FirstOrDefault(n => n.Key == "ValidImpactTypes");
+						var invalidImpactTypes = node.Value.Nodes.FirstOrDefault(n => n.Key == "InvalidImpactTypes");
+						if (validImpactTypes != null)
+						{
+							Console.WriteLine("CreateEffectWarhead now uses Valid-/InvalidTargets instead of Valid-/InvalidImpactTypes.");
+							Console.WriteLine("Please check whether you need to make manual adjustments.");
+
+							var validTargets = new List<string>();
+							if (validImpactTypes.Value.Value.Contains("Ground"))
+								validTargets.Add("Ground");
+							if (validImpactTypes.Value.Value.Contains("Water"))
+								validTargets.Add("Water");
+							if (validImpactTypes.Value.Value.Contains("Air"))
+								validTargets.Add("Air");
+
+							// 'validTargets' can be 0 here if the only valid ImpactType(s) were None, TargetHit or TargetTerrain.
+							// In that case we remove it and let the modder fix it manually.
+							if (validTargets.Count > 0)
+							{
+								validImpactTypes.Value.Value = validTargets.JoinWith(", ");
+								RenameNodeKey(validImpactTypes, "ValidTargets");
+							}
+							else
+								node.Value.Nodes.Remove(validImpactTypes);
+						}
+						else
+						{
+							// 'Air' is not part of the internal warhead ValidTargets default, but was part of the ValidImpactTypes default.
+							node.Value.Nodes.Add(new MiniYamlNode("ValidTargets", "Ground, Water, Air"));
```

Comment:
Normally it shouldn't, since ValidTargets previously had no effect for CEWH, but I'll add a safety-net just in case.